x/asset/types/msgs.go

```diff
@@ -0,0 +1,277 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var _ sdk.Msg = &MsgCreateAsset{}
+var _ sdk.Msg = &MsgCreateAssetClass{}
+var _ sdk.Msg = &MsgCreatePool{}
+var _ sdk.Msg = &MsgCreateTokenization{}
+var _ sdk.Msg = &MsgCreateSecuritization{}
+
+// AllRequestMsgs defines all the Msg*Request messages.
+var AllRequestMsgs = []sdk.Msg{
+	(*MsgCreateAsset)(nil),
+	(*MsgCreateAssetClass)(nil),
+	(*MsgCreatePool)(nil),
+	(*MsgCreateTokenization)(nil),
+	(*MsgCreateSecuritization)(nil),
+}
+
+// ValidateBasic implements Msg
+func (msg MsgCreateAsset) ValidateBasic() error {
+	if msg.Asset == nil {
+		return fmt.Errorf("asset cannot be nil")
+	}
+
+	if msg.Asset.ClassId == "" {
+		return fmt.Errorf("class_id cannot be empty")
+	}
+
+	if msg.Asset.Id == "" {
+		return fmt.Errorf("id cannot be empty")
+	}
+
+	if msg.Asset.Data != "" {
+		return validateJSON(msg.Asset.Data)
+	}
```

Comment:
issue (blocking): You need to only call `return` here if `validateJSON` returns an error.

As you have it, the `msg.FromAddress` (or anything checked after this) won't be checked unless there is no `msg.Asset.Data`. The `validateJSON` func handles the case where the data is empty, so you don't need to check that up-front.

```suggestion
	if err := validateJSON(msg.Asset.Data); err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}
```

Reply:
:heavy_check_mark: